### PR TITLE
Portability fixes

### PR DESCRIPTION
**Pull request by Artturin**
Better errors if something is missing

Not all platforms need 'dl' so use `CMAKE_DL_LIBS` which is set depending on
if its needed

Find threads with in a platform agnostic way which requires bumping
`cmake_minimum_required` to 3.10.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -75,7 +70,22 @@ ENDIF()
 
 
 find_program(SED NAMES gsed sed) # gsed - MacOS and FreeBSD, sed - all the rest
-find_file(PCRE_H pcre.h)
+
+pkg_search_module(PCRE REQUIRED libpcre)
+pkg_search_module(JSONC REQUIRED json-c)
+pkg_search_module(LIBGC REQUIRED bdw-gc-threaded bdw-gc)
```
Comment (ilyash-b) [inline CMakeLists.txt:76]: I never tested `bdw-gc`. Did you? If not, let's remove it.

File: CMakeLists.txt [hunk 2]
```diff
@@ -75,7 +70,22 @@ ENDIF()
...
+
+include_directories(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR} ${LIBFFI_INCLUDE_DIRS} /usr/local/include /opt/homebrew/include)
+link_directories(/usr/local/lib /opt/homebrew/lib)
+
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+find_package(Threads REQUIRED)
+
+file(READ lib/stdlib.ngs FILE_CONTENTS)
+string(REPLACE "#@INSTALL_LIBDIR@" ${CMAKE_INSTALL_LIBDIR}/ngs FILE_CONTENTS "${FILE_CONTENTS}")
+file(WRITE lib/stdlib.ngs "${FILE_CONTENTS}")
```
Comment (ilyash-b) [inline CMakeLists.txt:87]: I would really prefer passing `CMAKE_INSTALL_LIBDIR` to the C code and then exposing it into the language instead of this.

**Review by ilyash-b — COMMENTED** [review]
I get the following error on Mac:

```
[ 64%] Building C object CMakeFiles/ngs.dir/compile.c.o
[ 71%] Building C object CMakeFiles/ngs.dir/debug.c.o
[ 78%] Building C object CMakeFiles/ngs.dir/ast.c.o
[ 85%] Building C object CMakeFiles/ngs.dir/malloc.c.o
[ 92%] Building C object CMakeFiles/ngs.dir/decompile.c.o
[100%] Linking C executable ngs
ld: library 'gc' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [ngs] Error 1
make[2]: *** [CMakeFiles/ngs.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [build] Error 2
```

I don't know why this happens.

**Comment by ilyash-b** [comment]
While working on another issue discovered that this also happens when I comment out `link_directories(/usr/local/lib /opt/homebrew/lib)` in `CMakeLists.txt`

**Comment by Artturin** [comment]
I'll split out the dependency part

**Comment by ilyash-b** [comment]
Worked on https://github.com/ngs-lang/ngs/issues/668 and had to make some changes. I took some of your changes here to the `dev` branch. Please rebase.

**Comment by ilyash-b** [comment]
I have pulled your changes from this PR piece by piece into `dev`. Can you please take a look if I missed something? If everything is OK, you can close this PR.

I've modified string replacement to passing `CMAKE_INSTALL_FULL_LIBDIR` to the C code and from there to `NGS_PATH`.

**Comment by Artturin** [comment]
> I have pulled your changes from this PR piece by piece into `dev`. Can you please take a look if I missed something? If everything is OK, you can close this PR.
> 
> I've modified string replacement to passing `CMAKE_INSTALL_FULL_LIBDIR` to the C code and from there to `NGS_PATH`.

Awesome, thank you. There is one change missing. https://github.com/ngs-lang/ngs/pull/671

File: CMakeLists.txt (hunk 1 above)
Comment (Artturin) [reply CMakeLists.txt:76]: `bdw-gc` builds with threading by default, only freebsd separates the threaded and non threaded versions.

**Pull request closed without merging**